### PR TITLE
Add test for kube-scheduler command setup

### LINKED ISSUE #88028: Scheduler: Add tests for server.go for close-to-full E2E setup
We are lacking tests that exercise a full scheduler setup, all the way from config files and command line flags to what plugins end up instantiated.

The best candidate for this is `cmd/kube-scheduler/app/server.go`, but this needs tweaking. Basically, we need to split `runCommand` in two functions: `setupCommand` and the already existing  `Run` (which already has a test). Maybe we can have the test in `cmd/kube-scheduler/app/testing` as well.

Configuration can be written to files just like we do in `cmd/kube-scheduler/app/options/options_test.go` and args can be passed as parameters.

/sig scheduling
/help
/kind cleanup

### PR DESCRIPTION
**Pull request by notpad**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

We are lacking tests that exercise a full scheduler setup, all the way from config files and command line flags to what plugins end up instantiated. 

This PR does two things:
1. split `runCommand` function in  `cmd/kube-scheduler/app/server.go` into two parts, `setupCommand` and `Run`
2. add test for `setupCommand` to get a full scheduler setup from config files and command line flags, and check the instantiated plugins

**Which issue(s) this PR fixes**:

Fixes #88028

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

/sig-scheduling
/cc @alculquicondor

### CODE REVIEW

File: cmd/kube-scheduler/app/testing/command_test.go [hunk 1]
```diff
@@ -0,0 +1,361 @@
+package testing
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/testing/command_test.go]: Sorry, I was wrong. This should go to `cmd/kube-scheduler/app/server_test.go`. And `Run` doesn't have a test. It's instead being used in some integration tests.

This package is more of a library for other tests.

File: cmd/kube-scheduler/app/server.go [hunk 2]
```diff
@@ -339,3 +309,46 @@ func WithPlugin(name string, factory framework.PluginFactory) Option {
 		return registry.Register(name, factory)
 	}
 }
+
+// SetupCommand creates a completed config based on the command args and options
+func SetupCommand(cmd *cobra.Command, args []string, opts *options.Options) (*schedulerserverconfig.CompletedConfig, error) {
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server.go]: don't export this.

File: cmd/kube-scheduler/app/testing/command_test.go [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+			}
+			defer cc.SecureServing.Listener.Close()
+			defer cc.InsecureServing.Listener.Close()
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			recorderFactory := profile.NewRecorderFactory(events.NewBroadcaster(&events.EventSinkImpl{Interface: cc.EventClient.Events("")}))
+			outOfTreeRegistry := make(framework.Registry)
+			// Create the scheduler.
+			sched, err := scheduler.New(cc.Client,
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/testing/command_test.go]: This shouldn't be calling `scheduler.New`. Instead, the part of `app.Run` that calls `scheduler.New` should be part of `setupCommand`. And `Run` should do exactly that: call `sched.Run`.

File: cmd/kube-scheduler/app/testing/command_test.go (hunk 1 above)
Comment (notpad) [reply cmd/kube-scheduler/app/testing/command_test.go]: Done

File: cmd/kube-scheduler/app/server.go (hunk 2 above)
Comment (notpad) [reply cmd/kube-scheduler/app/server.go]: but `SetupCommand` can be used by `cmd/kube-scheduler/app/testing/testserver.go`, otherwise we need to create the scheduler manually in `cmd/kube-scheduler/app/testing/testserver.go`.

File: cmd/kube-scheduler/app/server.go [hunk 4]
```diff
@@ -339,3 +285,70 @@ func WithPlugin(name string, factory framework.PluginFactory) Option {
 		return registry.Register(name, factory)
 	}
 }
+
+// SetupCommand creates a completed config and a scheduler based on the command args and options
+func SetupCommand(ctx context.Context, args []string, opts *options.Options, outOfTreeRegistryOptions ...Option) (*schedulerserverconfig.CompletedConfig, *scheduler.Scheduler, error) {
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server.go]: nit: `Setup`, as this is not related to `cobra.Command` anymore.

File: cmd/kube-scheduler/app/testing/testserver.go [hunk 5]
```diff
@@ -111,14 +113,16 @@ func StartTestServer(t Logger, customFlags []string) (result TestServer, err err
 		s.CombinedInsecureServing.Metrics.Listener = listener
 		t.Logf("kube-scheduler will listen insecurely on port %d...", s.CombinedInsecureServing.BindPort)
 	}
-	config, err := s.Config()
+
+	cc, sched, err := app.SetupCommand(ctx, []string{}, s)
 	if err != nil {
 		return result, fmt.Errorf("failed to create config from options: %v", err)
 	}
+	config := cc.Config
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/testing/testserver.go]: having both `cc` and `config` is confusing. Prefer to use cc in the remaining lines.
The only exception is probably `result.Config = cc.Config`

File: cmd/kube-scheduler/app/testing/testserver.go [hunk 6]
```diff
@@ -66,6 +67,7 @@ func StartTestServer(t Logger, customFlags []string) (result TestServer, err err
 		if len(result.TmpDir) != 0 {
 			os.RemoveAll(result.TmpDir)
 		}
+		configz.Delete("componentconfig")
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/testing/testserver.go]: Where is this coming from?

File: cmd/kube-scheduler/app/server_test.go [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+						"QueueSortPlugin": {{Name: "PrioritySort"}},
+					},
+				},
+			},
+		},
+		{
+			name: "Deprecated SchedulerName flag",
+			flags: []string{
+				"--kubeconfig", configKubeconfig,
+				"--scheduler-name", schedulerName,
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server_test.go]: Are there any non-deprecated flags that could be easy to verify?

File: cmd/kube-scheduler/app/server_test.go [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+	kubeschedulerconfig "k8s.io/kubernetes/pkg/scheduler/apis/config"
+	"k8s.io/kubernetes/pkg/util/configz"
+)
+
+type schedulerProfile struct {
+	SchedulerName string
+	Plugins       map[string][]kubeschedulerconfig.Plugin
+}
+
+func TestCommand(t *testing.T) {
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server_test.go]: `TestSetup`

File: cmd/kube-scheduler/app/server_test.go [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+`, server.URL)), os.FileMode(0600)); err != nil {
+		t.Fatal(err)
+	}
+
+	v1alpha1Config := filepath.Join(tmpDir, "kubeconfig_v1alpha1.yaml")
+	schedulerName := "my-old-scheduler"
+	if err := ioutil.WriteFile(v1alpha1Config, []byte(fmt.Sprintf(`
+apiVersion: kubescheduler.config.k8s.io/v1alpha1
+kind: KubeSchedulerConfiguration
+schedulerName: "%s"
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server_test.go]: Just inline the name.

File: cmd/kube-scheduler/app/server_test.go [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+		},
+		{
+			name: "v1alpha1 config with SchedulerName and HardPodAffinitySymmetricWeight",
+			flags: []string{
+				"--config", v1alpha1Config,
+				"--kubeconfig", configKubeconfig,
+			},
+			wantProfiles: []schedulerProfile{
+				{
+					SchedulerName: schedulerName,
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server_test.go]: prefer to inline constants. It makes test tables easier to follow. But I think `defaultPlugins` is fine.

File: cmd/kube-scheduler/app/server_test.go [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+
+	testcases := []struct {
+		name         string
+		flags        []string
+		wantProfiles []schedulerProfile
+	}{
+		{
+			name: "config file",
+			flags: []string{
+				"--config", configFile,
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server_test.go]: This value is useless in this case, isn't it?

**Review by alculquicondor — COMMENTED** [review]
This is looking great. Thanks a lot

File: cmd/kube-scheduler/app/server.go (hunk 2 above)
Comment (alculquicondor) [reply cmd/kube-scheduler/app/server.go]: makes sense

File: cmd/kube-scheduler/app/server_test.go [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+			gotProfiles := make([]schedulerProfile, 0)
+			for n, p := range sched.Profiles {
+				profile := schedulerProfile{
+					SchedulerName: n,
+					Plugins:       p.ListPlugins(),
+				}
+				gotProfiles = append(gotProfiles, profile)
+			}
+
+			sortProfiles(tc.wantProfiles)
```
[the remark this answers is not in the document]
Comment (notpad) [reply cmd/kube-scheduler/app/server_test.go]: good suggestion

File: cmd/kube-scheduler/app/server.go [hunk 13]
```diff
@@ -119,79 +119,25 @@ func runCommand(cmd *cobra.Command, args []string, opts *options.Options, regist
...
-
-	// Get the completed config
-	cc := c.Complete()
-
-	// Configz registration.
-	if cz, err := configz.New("componentconfig"); err == nil {
-		cz.Set(cc.ComponentConfig)
-	} else {
-		return fmt.Errorf("unable to register configz: %s", err)
+	if cc == nil {
```
[the remark this answers is not in the document]
Comment (notpad) [reply cmd/kube-scheduler/app/server.go]: yes, for `--wirte-to-config`, cc will be nil

File: cmd/kube-scheduler/app/server.go (hunk 4 above)
Comment (notpad) [reply cmd/kube-scheduler/app/server.go]: done

File: cmd/kube-scheduler/app/server_test.go (hunk 8 above)
Comment (notpad) [reply cmd/kube-scheduler/app/server_test.go]: done

File: cmd/kube-scheduler/app/server_test.go (hunk 7 above)
Comment (notpad) [reply cmd/kube-scheduler/app/server_test.go]: added test for `write-to-config` flag

Comment (notpad) [reply cmd/kube-scheduler/app/server_test.go]: we can also add a test for the feature flag, but can we do that in a follow up PR? I find it is a little tricky to restore feature flags for different test cases.

File: cmd/kube-scheduler/app/testing/testserver.go (hunk 6 above)
Comment (notpad) [reply cmd/kube-scheduler/app/testing/testserver.go]: because we call `app.Setup` now, and it creates a configz, we need to clean up this config for next test case, or just log the error and don't return directly

```
	if cz, err := configz.New("componentconfig"); err == nil {
		cz.Set(cc.ComponentConfig)
	} else {
		return nil, nil, fmt.Errorf("unable to register configz: %s", err)
	}
```

File: cmd/kube-scheduler/app/server.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply cmd/kube-scheduler/app/server.go]: I feel like this should be part of `Run` instead of `Setup`. But that would require more refactoring.

File: cmd/kube-scheduler/app/server.go [hunk 14]
```diff
@@ -339,3 +285,70 @@ func WithPlugin(name string, factory framework.PluginFactory) Option {
...
+	if errs := opts.Validate(); len(errs) > 0 {
+		return nil, nil, utilerrors.NewAggregate(errs)
+	}
+
+	if len(opts.WriteConfigTo) > 0 {
+		c := &schedulerserverconfig.Config{}
+		if err := opts.ApplyTo(c); err != nil {
+			return nil, nil, err
+		}
+		if err := options.WriteConfigFile(opts.WriteConfigTo, &c.ComponentConfig); err != nil {
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server.go]: This specific block could you to `Run`. And maybe we shouldn't be calling `ApplyTo`, but just let the Setup go through all the other steps (like `opts.Config`). `write-to-config` is not intended for production, just for our own debugging. So the closer it is to a real run, the better.

File: cmd/kube-scheduler/app/server.go [hunk 15]
```diff
@@ -339,3 +285,70 @@ func WithPlugin(name string, factory framework.PluginFactory) Option {
...
+	c, err := opts.Config()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Get the completed config
+	cc := c.Complete()
+
+	// Configz registration.
+	if cz, err := configz.New("componentconfig"); err == nil {
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server.go]: Maybe part of `Run` as well?

File: cmd/kube-scheduler/app/testing/testserver.go (hunk 6 above)
Comment (alculquicondor) [reply cmd/kube-scheduler/app/testing/testserver.go]: I think we need to reach to the OWNERS of the users of this library (`/pkg/integration/serving`) to learn what are their expectations. Perhaps we can change this to be much closer to `runCommand`. Otherwise, anything that is not common, should go outside of `Setup`.

**Review by alculquicondor — COMMENTED** [review]
The PR is getting big and there are too many open questions. Let's leave it for after the freeze, given that it's not critical.
Still, I'd love to see this cleaned up soon.

File: cmd/kube-scheduler/app/server.go (hunk 15 above)
Comment (notpad) [reply cmd/kube-scheduler/app/server.go]: yes, we can, but I think this is more like a config "setup"

File: cmd/kube-scheduler/app/testing/testserver.go (hunk 6 above)
Comment (notpad) [reply cmd/kube-scheduler/app/testing/testserver.go]: Actually, if we move this to `Run`, we still need to add `configz.Delete("componentconfig")` because `app.Run` is also called by `StartTestServer`.

We can find similar code in `cmd/kube-controller-manager/app/controllermanager.go` which also has a testserver used by `/pkg/integration/serving`

```
func Run(c *config.CompletedConfig, stopCh <-chan struct{}) error {
	// To help debugging, immediately log version
	klog.Infof("Version: %+v", version.Get())

	if cfgz, err := configz.New(ConfigzName); err == nil {
		cfgz.Set(c.ComponentConfig)
	} else {
		klog.Errorf("unable to register configz: %v", err)
	}
```

it logs an error and continue the execution, otherwise it will also need a `configz.Delete`

File: cmd/kube-scheduler/app/server.go (hunk 14 above)
Comment (notpad) [reply cmd/kube-scheduler/app/server.go]: Do u mean move this block to Run? Maybe this flag is also useful to debug config problem in production env. Moving this to Run will make some differences, for example, the scheduler instance will be created, let me check if there are other unexpected effects.

Comment (alculquicondor) [reply cmd/kube-scheduler/app/server.go]: Specially when used in production, you want to see the configuration that is actually going to run. I guess the only effects we want to avoid is that other files get created or routines are started.

File: cmd/kube-scheduler/app/server.go (hunk 15 above)
Comment (alculquicondor) [reply cmd/kube-scheduler/app/server.go]: Well, this registers the component config in the /configz server. It comes down to the question in https://github.com/kubernetes/kubernetes/pull/88728#discussion_r387746862

File: cmd/kube-scheduler/app/testing/testserver.go (hunk 6 above)
Comment (alculquicondor) [reply cmd/kube-scheduler/app/testing/testserver.go]: That is true. But at least we don't need to worry about it when testing Setup. If this is what other binaries are doing for the serving test, that's great.

File: cmd/kube-scheduler/app/server.go (hunk 15 above)
Comment (notpad) [reply cmd/kube-scheduler/app/server.go]: moved to `Run`

File: cmd/kube-scheduler/app/server.go (hunk 14 above)
Comment (notpad) [reply cmd/kube-scheduler/app/server.go]: moved to `Run`

File: cmd/kube-scheduler/app/server.go [hunk 16]
```diff
@@ -126,18 +126,30 @@ func runCommand(cmd *cobra.Command, args []string, opts *options.Options, regist
...
-	}
 
-	return Run(ctx, cc, sched)
+	return Run(ctx, cc, opts, sched)
 }
 
 // Run executes the scheduler based on the given configuration. It only returns on error or when context is done.
-func Run(ctx context.Context, cc *schedulerserverconfig.CompletedConfig, sched *scheduler.Scheduler) error {
+func Run(ctx context.Context, cc *schedulerserverconfig.CompletedConfig, opts *options.Options, sched *scheduler.Scheduler) error {
+	if len(opts.WriteConfigTo) > 0 {
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/server.go]: Sorry for the back-and-forth. I think we can have this directly in `runCommand` instead, right before calling Run. Otherwise, there is no other reason to pass both `cc` and `opts` to run. And we don't need to WriteConfigTo in unit tests or serving tests.

**Review by alculquicondor — COMMENTED** [review]
looking good. There is a rebase needed.

File: cmd/kube-scheduler/app/server.go (hunk 16 above)
Comment (notpad) [reply cmd/kube-scheduler/app/server.go]: done

File: cmd/kube-scheduler/app/testing/testserver.go [hunk 17]
```diff
@@ -111,20 +113,21 @@ func StartTestServer(t Logger, customFlags []string) (result TestServer, err err
 		s.CombinedInsecureServing.Metrics.Listener = listener
```
Comment (alculquicondor) [inline cmd/kube-scheduler/app/testing/testserver.go]: nit: rename s to opts

**Review by alculquicondor — COMMENTED** [review]
You can squash.

Although, it seems that master is still under freeze.

**Comment by notpad** [comment]
@alculquicondor code is ready for final review

File: cmd/kube-scheduler/app/server_test.go [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+				t.Fatal(err)
+			}
+			for _, f := range opts.Flags().FlagSets {
+				fs.AddFlagSet(f)
+			}
+			if err := fs.Parse(tc.flags); err != nil {
+				t.Fatal(err)
+			}
+
+			args := []string{}
```
Comment (Huang-Wei) [inline cmd/kube-scheduler/app/server_test.go]:
```suggestion
			var args []string
```

**Review by Huang-Wei — COMMENTED** [review]
Just one nit. Feel free to cancel the /hold after addressing the nit.

/lgtm
/hold
